File: examples/wc.cpp
#include <proton/base.hpp>
#include <proton/getopt.hpp>
#include <proton/io.hpp>
#include <proton/string.hpp>

using namespace std;
using namespace proton;

void usage()
{
    cout <<
"Usage: wc [OPTION]... [FILE]...\n"
"  -c, --bytes            print the byte counts\n"
"  -l, --lines            print the newline counts\n"
"  -w, --words            print the word counts\n"
"      --help     display this help and exit\n"
    << endl;
}

tuple<long, long, long> count_file(const str& fn, bool cb, bool cw, bool cl)
{
    long total_bytes=0, total_words=0, total_lines=0;

    fstream f(fn.c_str());
    str line;
    while(len(line=readline(f))){
        total_lines++;
        auto words=line.split();
        total_words+=len(words);
        total_bytes=f.tellg();
    }
    return _t(total_bytes,total_words,total_lines);
}

int main(int argc, char** argv)
{
    getopt_t g;
    try{
        g=getopt(argc, argv, "clw", {"bytes","lines","words","help"});
    }
    catch(invalid_argument& e){
        usage();
        return -1;
    }

    auto opts=at<0>(g);
    auto args=at<1>(g);

    bool count_bytes=true;
    bool count_lines=true;
    bool count_words=true;

    if(len(opts)>0){
        count_bytes=false;
        count_lines=false;
        count_words=false;
    }

    for(auto x : opts){
        auto s=at<0>(x);
        if(s=="-c" || s=="--bytes"){
            count_bytes=true;
        }
        else if(s=="-w" || s=="--words"){
            count_words=true;
        }
        else if(s=="-l" || s=="--lines"){
            count_lines=true;
        }
        else if(s=="--help"){
            usage();
            return 0;
        }
    }

    for(auto x:args){
        long bytes, words, lines;
        tie(bytes, words, lines)=count_file(x, count_bytes, count_words, count_lines);
        cout << " " ;
        if(count_lines)
            cout << lines << " ";
        if(count_words)
            cout << words << " ";
        if(count_bytes)
            cout << bytes << " ";
        cout << x << endl;
    }

    return 0;
}

File: examples/wc.py
import getopt, sys

def count_file(fn, cb, cw, cl):
    total_bytes=0
    total_words=0
    total_lines=0

    f=file(fn)
    for line in f:
        total_lines+=1
        words=line.split()
        total_words+=len(words)
        total_bytes=f.tell()
    return total_bytes, total_words, total_lines

try:
    opts,args=getopt.getopt(sys.argv[1:], "cwl", ["bytes","lines","words","help"])
except:
    print "bad arg"
    sys.exit(-1)

count_bytes=True
count_lines=True
count_words=True

if len(opts)>0:
    count_bytes=False
    count_lines=False
    count_words=False

for x in opts:
    s=x[0]
    if(s=='-c' or s=='--bytes'):
        count_bytes=True
    elif (s=='-w' or s=='--words'):
        count_words=True
    elif (s=='-l' or s=='--lines'):
        count_lines=True
    elif (s=='--help'):
        print 'help'
        sys.exit(0)

for x in args:
    bytes, words, lines=count_file(x, count_bytes, count_words, count_lines)
    if count_lines:
        print lines,
    if count_words:
        print words,
    if count_bytes:
        print bytes,
    print x

sys.exit(0)
